### ISSUE TITLE
feat: gbrain skillopt — SkillOpt-style self-evolving agent skills

### ISSUE THREAD

**Issue by garrytan-agents** [garrytan/gbrain#1481]
## Summary

Proposal to add `gbrain skillopt` — a validation-gated, bounded text-space optimizer for agent skill documents. Based on [SkillOpt](https://arxiv.org/abs/2605.23904) (Microsoft Research, May 2026), which treats the skill doc as trainable external state of a frozen agent with deep-learning-style controls.

SkillOpt is best or tied on all 52 evaluated (model, benchmark, harness) cells. It adds +23.5 points over no-skill on GPT-5.5, beats human-written skills, TextGrad, GEPA, EvoSkill, and Trace2Skill, and the learned skills transfer across models and harnesses.

gbrain already has: 47 bundled skills (SKILL.md), routing-eval fixtures, `gbrain eval --qrels`, `gbrain dream`, and `gbrain skillify`. What's missing is closing the loop — skills don't evolve from execution feedback.

---

### User needs
- Skills are hand-written and static. Once shipped, they only improve when a human rewrites them.
- Human-written skills plateau because humans can't process 40+ execution trajectories to find which rules help and which hurt.
- Agents using skills accumulate failure patterns that the skill never adapts to.
- Different target models (GPT-5.5, Claude, Gemini) need different skill formulations, but currently every model gets the same SKILL.md.

### Agent needs
- Skills should improve from production feedback without human intervention.
- The improvement loop must be safe — never ship a skill that's worse than what it replaced.
- Skills must remain inspectable (markdown, not opaque weights).
- Optimization cost must be bounded and amortizable (train once, deploy everywhere).
- The skill format shouldn't change — existing SKILL.md files should be valid inputs.

---

## Prior Art Landscape

| System | What it optimizes | Feedback | Controls | Skills transfer? |
|--------|------------------|----------|----------|-----------------|
| APE (2022) | Prompts | Generate + select | None | No |
| OPRO (2023) | Prompts | Meta-prompt with scores | None | No |
| DSPy (2023+) | Pipeline parameters | Module-level metrics | Compilation-time | No |
| TextGrad (Nature 2025) | Any text variable | Textual gradients | PyTorch-like API | No |
| GEPA (2025) | Prompts + code | Execution traces + Pareto | Genetic selection | Partial |
| EvoSkill (Apr 2026) | Agent skills | Co-evolutionary verification | Mutation + selection | Cross-model |
| Trace2Skill (Mar 2026) | Skills from trajectories | Parallel fleet analysis | Hierarchical patches | Cross-model |
| **SkillOpt (May 2026)** | **Agent skill docs** | **Scored rollouts** | **LR, validation gate, rejected-edit buffer, slow/meta update** | **Cross-model, cross-harness, cross-benchmark** |

SkillOpt's key insight: apply the same discipline that makes weight-space optimization reproducible — learning rate, validation gating, negative feedback memory — to text-space optimization.

### Implementations reviewed
- [microsoft/SkillOpt](https://github.com/microsoft/SkillOpt) — official, Python, full pipeline
- [zou-group/textgrad](https://github.com/zou-group/textgrad) — Nature-published text optimization
- [stanfordnlp/dspy](https://github.com/stanfordnlp/dspy) — programmatic prompt compilation
- [google-deepmind/opro](https://github.com/google-deepmind/opro) — LLM-as-optimizer
- [gepa-ai/gepa](https://github.com/gepa-ai/gepa) — genetic-pareto with traces
- [sentient-agi/EvoSkill](https://github.com/sentient-agi/EvoSkill) — co-evolutionary skill discovery
- [AMAP-ML/SkillClaw](https://github.com/AMAP-ML/SkillClaw) — skill evolution from sessions
- [NousResearch/hermes-agent-self-evolution](https://github.com/NousResearch/hermes-agent-self-evolution) — DSPy + GEPA

---

## The SkillOpt Algorithm

```
Input: initial skill s₀, target model M, harness h,
       D_train, D_sel, D_test, num_epochs, batch_size,
       learning rate schedule L_t

s_best = s₀
best_score = eval(s₀, D_sel)
rejected_buffer = []

for epoch in 1..num_epochs:
    for step in 1..steps_per_epoch:
        # 1. FORWARD PASS: rollout batch
        trajectories = [h(M, x, s_current) for x in sample(D_train)]
        
        # 2. BACKWARD PASS: minibatch reflection
        fail_edits = reflect(failures, s_current, rejected_buffer)
        succ_edits = reflect(successes, s_current, rejected_buffer)
        
        # 3. BOUNDED UPDATE: apply under learning rate
        edits = rank_and_clip(merge(fail_edits, succ_edits), L_t)
        s_candidate = apply_edits(s_current, edits)
        
        # 4. VALIDATION GATE: accept only if improved
        if eval(s_candidate, D_sel) > best_score:
            s_current = s_candidate
            best_score = eval(s_candidate, D_sel)
        else:
            rejected_buffer.append(edits)
    
    # 5. EPOCH-WISE SLOW UPDATE
    s_current = apply_slow_update(s_current, compare_epochs(...))

export(s_best)  # 300-2000 token skill document
```

### The Deep Learning Analogy (Operational, Not Decorative)

| DL Concept | SkillOpt Analog |
|------------|-----------------|
| Parameters | Skill document (natural language) |
| Forward pass | Rollout batch (execute tasks with current skill) |
| Backward pass | Minibatch reflection (analyze trajectories) |
| Gradient | Proposed add/delete/replace edits |
| Learning rate | Edit budget L_t (max edits per step) |
| LR schedule | Cosine/linear/constant decay |
| Validation set | D_sel (held-out gate) |
| Momentum | Epoch-wise slow update |
| Negative examples | Rejected-edit buffer |
| Checkpointing | best_skill.md |

### Ablation Results (what matters)

| Component removed | Effect |
|-------------------|--------|
| Learning rate (bounded edits) | Unbounded rewriting erases useful rules |
| Validation gate | Harmful proposals accumulate |
| Rejected-edit buffer | Optimizer repeats failed edits |
| Slow/meta update | Long-horizon patterns lost |
| Cosine schedule | Slight degradation vs cosine |

---

### CLI Interface

```bash
gbrain skillopt <skill-name> \
  --benchmark <path>          # Task set with expected outcomes
  --epochs 4                  # Training epochs (default: 4)
  --batch-size 40             # Rollout batch per step (default: 40)
  --lr 4                      # Max edits per step (default: 4)
  --lr-schedule cosine        # Edit budget schedule (default: cosine)
  --optimizer-model opus      # Optimizer model (default: models.think fallback)
  --target-model default      # Target model executing the skill
  --split 4:1:5               # train:validation:test ratio
  --patch                     # Patch mode (default) vs --rewrite
  --dry-run                   # Preview optimization plan without executing
  --json                      # Machine-readable output
```

### Benchmark Format

Extend the existing `routing-eval.jsonl` with execution expectations:

```jsonl
{"task": "Research company Acme Corp for an investor meeting", "skill": "meeting-prep", "expected_elements": ["company_profile", "key_people", "recent_news"], "judge": "llm"}
{"task": "Draft a tweet about network effects", "skill": "tweet-composition", "expected_elements": ["claim", "evidence", "under_280_chars"], "judge": "rule"}
{"task": "Check if claim X is supported by the brain", "skill": "fact-check", "expected_elements": ["verdict", "citations", "confidence"], "judge": "llm"}
```

### Scoring Functions

Three built-in scoring modes:

1. **Rule-based** (`judge: rule`): Check for expected structural elements (has citations? under length limit? includes required sections?)
2. **LLM-as-judge** (`judge: llm`): Optimizer model scores execution quality 0-1 against the expected outcome
3. **Qrels** (`judge: qrels`): Existing `gbrain eval` retrieval metrics for search/retrieval skills

### Edit Operations

```yaml
# Example edit proposal from optimizer
edits:
  - op: add
    location: "## Error Handling"
    content: "When gbrain think returns no citations, fall back to gbrain search with broader keywords before reporting 'not found'."
  - op: replace
    target: "Always use gbrain search for lookups"
    replacement: "Use gbrain think for multi-page synthesis. Fall back to gbrain search only for simple keyword lookups."
  - op: delete
    target: "Include a disclaimer about AI-generated content"
    reason: "This rule consistently causes output to include unnecessary boilerplate that lowers quality scores."
```

### Output Artifacts

```
skills/<skill-name>/
├── SKILL.md               ← updated best skill (or unchanged if no improvement)
├── skillopt/
│   ├── history.json        ← full optimization trace
│   ├── rejected.json       ← rejected-edit buffer (negative feedback)
│   ├── versions/
│   │   ├── v0000_initial.md
│   │   ├── v0001_step1.md
│   │   └── v0042_best.md
│   └── report.md           ← human-readable optimization report
```

### Integration with Existing Commands

| Command | Current | With SkillOpt |
|---------|---------|---------------|
| `gbrain skillify` | Creates new skills | Creates → then `skillopt` optimizes them |
| `gbrain eval` | Retrieval quality | Extended: skill execution quality |
| `gbrain dream` | Overnight maintenance | Optional `--phase skillopt` for skill optimization |
| `gbrain check-resolvable` | Routing health | Also reports skills with low eval scores |
| `gbrain doctor` | Brain health | Reports skill optimization status |

### Dream Cycle Integration

```
gbrain dream --phase skillopt
```

The dream cycle already runs overnight. Adding a `skillopt` phase:
1. For each skill with a benchmark file and an eval score below threshold
2. Run one epoch of SkillOpt (bounded by time/cost budget)
3. If improvement found, update SKILL.md and commit
4. Report in the CycleReport

This means skills improve every night, automatically, with the same validation discipline as the rest of the dream cycle.

---

### From the paper
1. **Optimizer model quality matters:** Needs frontier model for reflection. Weak optimizers produce worse skills.
2. **Training cost:** Each step = rollout batch + validation eval. At batch_size=40, this is significant. Budget controls are essential.
3. **Local optima:** Validation gating prevents regression but can't guarantee global optimum.
4. **Small training sets:** < 20% of data significantly degrades performance.
5. **Rewrite mode instability:** Full rewrites can erase useful rules. Patch mode is safer for incremental refinement.

### gbrain-specific
6. **Skills are router-triggered, not batch-executed.** SkillOpt assumes batch execution. gbrain needs an adapter that executes a skill on a task and captures the trajectory.
7. **No scalar scoring today.** `routing-eval.jsonl` tests routing accuracy, not execution quality. Need execution-level scoring (the benchmark format above).
8. **Multi-skill routing.** Optimizing one skill could affect routing for others. Need isolation: optimize one skill's execution quality while holding routing fixed.
9. **Skill size constraints.** SkillOpt outputs 300-2000 tokens. gbrain SKILL.md files are often 2000-5000 tokens. May need to separate the "trained" section from static reference material.
10. **Cold start for new skills.** `gbrain skillify` creates an initial skill. SkillOpt needs a benchmark to optimize against — who creates the benchmark?

### Mitigations
- **Cost control:** `--max-cost` flag, per-step cost tracking, dream phase has a time budget
- **Safety:** Validation gating is mandatory (not optional). Rejected-edit buffer prevents regression loops.
- **Routing isolation:** SkillOpt only modifies the skill's execution behavior, not its routing triggers.
- **Benchmark bootstrapping:** For skills with `routing-eval.jsonl`, auto-generate an execution benchmark from the routing fixtures. For new skills, require a benchmark file.

---

### 1. Overnight skill improvement
A brain with 47 skills runs `gbrain dream --phase skillopt`. Each skill with a benchmark gets one epoch of optimization. Over a week, skills that had 60% execution quality reach 85%+.

### 2. New skill onboarding
User runs `gbrain skillify "meeting prep"` to create a skill. Then `gbrain skillopt meeting-prep --benchmark meeting-prep-tasks.jsonl` to optimize it against real tasks. The skill goes from generic to domain-adapted in 4 epochs.

### 3. Model migration
User switches from GPT-5.4 to Claude 4.5. Runs `gbrain skillopt --all --target-model claude-4.5` to re-optimize all skills for the new model. Skills transfer but may need fine-tuning for model-specific patterns.

### 4. Skill debugging
A skill is underperforming. `gbrain skillopt <skill> --dry-run` analyzes the benchmark and reports: "This skill fails on tasks involving multi-page synthesis because it uses `gbrain search` instead of `gbrain think`." The optimization trace shows exactly which rules cause failures.

---

### Phase 1: Foundation (MVP)
- `gbrain skillopt` command with benchmark input
- Patch-mode edits with validation gating
- LLM-as-judge scoring
- Rejected-edit buffer
- Cosine LR schedule
- Output: updated SKILL.md + history.json

### Phase 2: Dream Integration
- `gbrain dream --phase skillopt`
- Auto-discovery of skills with benchmarks
- Per-skill time/cost budgets
- Reporting in CycleReport

### Phase 3: Ecosystem
- Auto-generate benchmarks from routing-eval.jsonl
- Cross-model optimization (optimize once, deploy to multiple models)
- Skill quality dashboard in admin UI
- Meta-skill (optimizer-side bookkeeping for long-horizon patterns)

---

## References

- [SkillOpt paper](https://arxiv.org/abs/2605.23904) — Yang et al., Microsoft Research, May 2026
- [microsoft/SkillOpt](https://github.com/microsoft/SkillOpt) — official implementation
- [TextGrad](https://github.com/zou-group/textgrad) — Nature 2025, textual gradient descent
- [DSPy](https://github.com/stanfordnlp/dspy) — programmatic prompt optimization
- [OPRO](https://github.com/google-deepmind/opro) — LLM-as-optimizer
- [GEPA](https://github.com/gepa-ai/gepa) — genetic-pareto execution trace analysis
- [EvoSkill](https://github.com/sentient-agi/EvoSkill) — co-evolutionary skill discovery
- [SkillsBench](https://arxiv.org/abs/2602.06056) — benchmark for agent skills
- [SoK: Agentic Skills](https://arxiv.org/abs/2602.20867) — skill lifecycle taxonomy
